gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GobblinServiceManager.java

```diff
@@ -191,6 +191,7 @@
 
   protected Optional<HelixLeaderState> helixLeaderGauges;
 
+  public D2Announcer d2Announcer;
```

Comment:
How is it initialized? If it is optional, and is going to be set through dependency injection, it needs  @Inject(optional = true) attribute.

Reply:
Right now I am just directly setting this field after the `GobblinServiceManager` is initialized.

How does the dependency injection work exactly? Like I see for the `FlowStatusGenerator` we call `binder.bind(FlowStatusGenerator.class);`, does that construct a `FlowStatusGenerator` and inject it, and if so how does it provide arguments to the constructor?

I needed to, in our internal code, construct a subclass of D2Announcer and provide it to `GobblinServiceManager`, I couldn't figure out how to do it in a different way.

Reply:
For li-specific implementations, you can change LiGobblinServiceGuiceModule to override the defaults and inject private implementations. If your class depends on other interfaces registered in Guice, they will be injected into its constructor. Search for "@Inject" annotation on constructors in other classes to see how it works.

if the subclass of D2Announcer depends on some instances that you don't create yourself,  you can register such instances in Guice like this - https://github.com/apache/gobblin/blob/b5e32b158bd713c45caa5b4b8866f8a4c4fc3a29/gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GobblinServiceGuiceModule.java#L104 , or create a provider for them/D2Announcer like this - https://github.com/apache/gobblin/blob/b5e32b158bd713c45caa5b4b8866f8a4c4fc3a29/gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GobblinServiceGuiceModule.java#L184

Here is more info on how Guice works - https://github.com/google/guice/wiki/Motivation

Reply:
Thanks for the explanation, was able to get it working through injection.